<a href='http://www.algebra.hr'> <img src='../algebra_logo_color_h.png' alt="Algebra" width="500" /></a>
___
# NumPy - Numerical Python

NumPy – Numerical Python neizostavni je paket za numeričke proračune u Pythonu. Alat naziva Matlab koristi se za tehničke proračune, kao na primjer proračun opterećenja čelične konstrukcije mosta ili toplinskog naprezanja u bloku motora tijekom rada i sl. Kako su povezani Python i alat Matlab? Matlab je snažan alat, ali u njemu nije tako jednostavno pisati kôd. Bilo bi odlično kada bismo imali snagu Matlaba, a jednostavnost Pythona. Tako dolazimo do seta biblioteka (skupina klasa i metoda određenih funkcionalnosti) koji koriste Python i koji su alternativa Matlabu.

U tom setu temelj je NumPy. Namijenjen je za numeričke proračune. Zatim je ovdje SciPy (*Scientific Python*) razvijen na temeljima NumPy biblioteke i nudi veliki izbor gotovih algoritama za razne znanstvene proračune. Tu je i Pandas, također razvijen na temeljima NumPy biblioteke, koji služi za obradu podataka u obliku tablice. Matplotlib biblioteka koristi NumPy kao temelj, a služi za vizualizaciju podataka.

Pandas i matplotlib smo već upoznali. Upoznajmo se s njihovom temeljnom bibliotekom NumPy.

NumPy je *open-source* biblioteka namijenjena za kompleksne matematičke proračune. Nudi jako puno matematičkih funkcija, generator je nasumičnih brojeva, podrška je za proračune vezane uz linearnu algebru, Fourierove transformacije. Mi ćemo ga koristiti za obradu višedimenzionalnih nizova i matrica.

Na službenoj stranici NumPy biblioteke možete pronaći dokumentaciju, preuzeti i instalirati NumPy, pronaći neke tečajeve, pridružiti se zajednici korisnika. Poveznica na stranicu je: https://numpy.org/.

## NumPy primjena

Već smo spomenuli da se NumPy najčešće koristi za znanstvene matematičke proračune, ali tu nije kraj primjeni NumPy paketa. Intenzivno se koristi i za statističke proračune, matematičke operacije nad matricama, kopiranje i pregled nizova, linearnu algebru, pretraživanje, sortiranje i prebrojavanje elemenata nizova i matrica.

# ndarray objekt - temelj NumPy biblioteke

Temelj NumPy biblioteke je *ndarray* objekt. Naziv dolazi od *n-dimensional array*, odnosno višedimenzionalni niz. Prije nego pojasnimo što su višedimenzionalni nizovi, dobro bi bilo objasniti što je to niz, odnosno *array*. *Array* je tip podatka koji je najsličniji Python listi. Nije isti jer postoje razlike, ne samo u načinu kako je niz programiran za rad s podacima (što je tema za kolegije na fakultetima), nego i nekoliko je nama "vidljivih" razlika. *Array* (mi ćemo dalje u tekstu ponekad koristiti i naziv *Array*):
- je kolekcija elemenata istog tipa podataka
- podržava izravne matematičke operacije nad elementima (*array* * 2 - množi svaki element niza s brojem dva)
- nije zamišljen za mijenjanje elemenata. podržava ih, ali je daleko manje učinkovit od liste.

Najbolje je napraviti nekoliko primjera pa će zasigurno biti jasnije.

Kao i uvijek do sada, prvo moramo napraviti *import* NumPy biblioteke kako bismo kasnije mogli koristiti dostupne mogućnosti.

Preporuka je koristiti način uključivanja biblioteke koji je gotovo postao standard. Uporaba drugačijeg načina neće utjecati na funkcionalnost programa, ali neće biti u skladu s najboljim praksama.

In [1]:
import numpy as np

In [2]:
python_lista = [1, 2, 3, 4, 5]
np_array = np.array([1, 2, 3, 4, 5])

print(f'Mnozenje python liste {python_lista * 2}')
print(f'Mnozenje NumPy matrice {np_array * 2}')

Mnozenje python liste [1, 2, 3, 4, 5, 1, 2, 3, 4, 5]
Mnozenje NumPy matrice [ 2  4  6  8 10]


Ali pomoću NumPy biblioteke Python listu, možemo pretvoriti u ndarray i pomnožiti svaki element s, recimo tri, jako jednostavno.

In [3]:
np.multiply(python_lista, 3)

array([ 3,  6,  9, 12, 15])

**PITANJE<br>
U kojoj situaciji metoda *multiply* neće raditi, odnosno javit će grešku?<br>
Pokažite na primjeru.**

In [4]:
python_lista = [1, 2, 3, 'Algebra', 'Python', 4, 5, 3.1415, False]
np.multiply(python_lista, 3)

UFuncTypeError: ufunc 'multiply' did not contain a loop with signature matching types (dtype('<U32'), dtype('int32')) -> None

Nastavimo dalje s istraživanjem osnovnih mogućnosti *ndarray* tipa podataka.

### *ndarray size* - broj elemenata u *Array* objektu

Svojstvo *size* vraća ukupan broj elemenata u objektu. Pokažimo to na primjeru.

In [7]:
podaci = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
print(podaci)
np_podaci = np.array(podaci)
print(np_podaci)

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]
[[1 2 3]
 [4 5 6]
 [7 8 9]]


Uočite kako je ispis podataka drugačiji.

Provjerimo koliko elemenata ima novokreirani *Array* objekt:

In [8]:
np_podaci.size

9

Vidjeli smo kako je ispis za *Array* drugačiji od obične Python liste. Svaku listu je prikazao u novom redu. Na ovaj način smo dobili 3x3 matricu.

Ako zamislimo da je NumPy *Array* u koordinatnom sustavu, svaki od redova bit će na X osima, svaka kolona će biti na Y osi. Ovaj podatak zovemo dimenzije, odnosno osi ili engleski *axis*. Jako često ćemo trebati informaciju koliko dimenzija ima naš *Array* objekt. Za to koristimo svojstvo ***ndim***. Provjerimo koliko naš *Array* objekt ima osi:

In [10]:
np_podaci.ndim

2

Očekivano, dvije osi. Kako izgleda *Array* s tri osi?

In [29]:
lista = [[[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]],
        [[13, 14, 15, 16], [17, 18, 19, 20], [21, 22, 23, 24]],
        [[25, 26, 27, 28], [29, 30, 31, 32], [33, 34, 35, 36]]]

tri_d = np.array(lista)
tri_d

array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]],

       [[25, 26, 27, 28],
        [29, 30, 31, 32],
        [33, 34, 35, 36]]])

In [16]:
tri_d.ndim

3

Dakle, treću dimenziju ćemo dobiti ukoliko *Array* kreiramo od liste, koja kao elemente ima listu, koja opet kao elemente ima listu.

A kakvog je oblika naš *Array*? Ako koristimo svojstvo *Shape*, dobit ćemo *tuple* s brojem redova i stupaca koje imamo, ali po svakoj dimenziji. Najbolje je to pokazati na primjeru kako bi vam bilo jasnije.

In [17]:
np_podaci.shape

(3, 3)

Vidimo da se naš *Array np_podaci* sastoji od dvije dimenzije (imamo samo dva podatka u *tupleu*) u kojima imamo tri reda i tri stupca. Provjerimo kakvog je oblika *Array tri_d*:

In [18]:
tri_d.shape

(3, 3, 4)

Sada smo dobili tri podatka pa je zaključak kako imamo **tri** dimenzije i u svakoj od dimenzija imamo **tri reda** i **četiri stupca**.

Već smo rekli kako u *Array* možemo pohraniti podatke **istog tipa**, ali to ne znači da ti podaci moraju biti isključivo cijeli brojevi. Osim toga, *Array* je namijenjen za rad s velikom količinom podatka pa je podataka o tipu i količini memorije koju taj tip zauzima jako koristan. Provjerimo koji je tip (*dtype*) u našim *Array* objektima te koliko memorije oni zauzimaju pojedinačno (*itemsize*), odnosno ukupno (*nbytes*).

In [19]:
print(np_podaci.nbytes)
print(np_podaci.itemsize)
print(np_podaci.dtype)

print(tri_d.nbytes)
print(tri_d.itemsize)
print(tri_d.dtype)

36
4
int32
144
4
int32


Vidimo da oba *Array* objekta imaju isti tip (*int64*). To možemo definirati tijekom kreiranja *Array* objekta tako da metodi *array* proslijedimo argument *dtype* s nazivom tipa podataka. Probajmo umjesto cijelih brojeva koristiti decimalne.

In [30]:
tri_d = np.array(tri_d, dtype=np.float64)
tri_d

array([[[ 1.,  2.,  3.,  4.],
        [ 5.,  6.,  7.,  8.],
        [ 9., 10., 11., 12.]],

       [[13., 14., 15., 16.],
        [17., 18., 19., 20.],
        [21., 22., 23., 24.]],

       [[25., 26., 27., 28.],
        [29., 30., 31., 32.],
        [33., 34., 35., 36.]]])

Provjerimo sada tip i količinu memorije za *Array tri_d*.

In [21]:
print(tri_d.nbytes)
print(tri_d.itemsize)
print(tri_d.dtype)

288
8
float64


Tip je sada drugačiji, a količina memorije je ostala ista jer *int64* i *float64* zauzimaju isto memorije (*8 byte*).

### Reshape

*Reshape* je metoda pomoću koje možemo promijeniti oblik, odnosno *shape* našeg *Array* objekta. Pokažimo to na primjeru. *Reshape* metoda prima *tuple* kao argument, ali umnožak njegovih elemenata mora biti isti kao i umnožak inicijalnih vrijednosti *Array* objekta. To znači da smo mi imali *tuple* (3, 3, 4) kao rezultat, a umnožak tih brojeva je 36. To znači da umnožak elemenata novog oblika mora biti 36. Recimo, (9, 4).

In [31]:
print(f'Shape tri_d prije: \t{tri_d.shape}')
print(f'tri_d prije: \n{tri_d}')

tri_d = tri_d.reshape(9, 4)

print(f'Shape tri_d poslije: \t{tri_d.shape}')
print(f'tri_d poslije: \n{tri_d}')

Shape tri_d prije: 	(3, 3, 4)
tri_d prije: 
[[[ 1.  2.  3.  4.]
  [ 5.  6.  7.  8.]
  [ 9. 10. 11. 12.]]

 [[13. 14. 15. 16.]
  [17. 18. 19. 20.]
  [21. 22. 23. 24.]]

 [[25. 26. 27. 28.]
  [29. 30. 31. 32.]
  [33. 34. 35. 36.]]]
Shape tri_d poslije: 	(9, 4)
tri_d poslije: 
[[ 1.  2.  3.  4.]
 [ 5.  6.  7.  8.]
 [ 9. 10. 11. 12.]
 [13. 14. 15. 16.]
 [17. 18. 19. 20.]
 [21. 22. 23. 24.]
 [25. 26. 27. 28.]
 [29. 30. 31. 32.]
 [33. 34. 35. 36.]]


Izgubili smo jednu dimenziju. Svedimo *tri_d Array* na samo jednu dimenziju.

In [34]:
tri_d = tri_d.reshape(36)
tri_d

array([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12., 13.,
       14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25., 26.,
       27., 28., 29., 30., 31., 32., 33., 34., 35., 36.])

Dakle, lista je jednodimenzionalni niz.

### Generator nasumičnih brojeva

Do sada smo vidjeli kako NumPy *ndarray* možemo kreirati iz neke Python liste ili izravno pomoću *ndarray()* metode. Međutim, NumPy nudi nekoliko načina za kreiranje *Array* objekata od nasumično kreiranih objekata koje možemo koristiti u našim matematičkim analizama.

Krenimo s generatorom *Arraya* objekta u kojem su sve nule, tako što ćemo koristiti metodu *zeros()*, kojoj ćemo kao argument predati *tuple* s definiranim oblikom (*shape*) željenog *Array* objekta.

In [41]:
nule = np.zeros((3, 4, 4))
nule

array([[[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]],

       [[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]],

       [[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]]])

Ako imamo metodu *zeros()*, moguće je da postoji identična metoda za jedinice naziva *ones()*. Provjerimo, ali pokušajmo promijeniti tip podatka u cijele brojeve (*dtype*).

In [42]:
jedinice = np.ones((3, 4, 4))
jedinice

array([[[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]],

       [[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]],

       [[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]]])

Odlično. Ipak, bilo bi daleko korisnije kada bismo mogli imati *Array* s predefiniranom vrijednošću, kao recimo 3.14. Već smo vidjeli da *Array* objekt možemo množiti pa pokušajmo to iskoristiti.

In [43]:
nd_pi = 3.1415 * jedinice
nd_pi

array([[[3.1415, 3.1415, 3.1415, 3.1415],
        [3.1415, 3.1415, 3.1415, 3.1415],
        [3.1415, 3.1415, 3.1415, 3.1415],
        [3.1415, 3.1415, 3.1415, 3.1415]],

       [[3.1415, 3.1415, 3.1415, 3.1415],
        [3.1415, 3.1415, 3.1415, 3.1415],
        [3.1415, 3.1415, 3.1415, 3.1415],
        [3.1415, 3.1415, 3.1415, 3.1415]],

       [[3.1415, 3.1415, 3.1415, 3.1415],
        [3.1415, 3.1415, 3.1415, 3.1415],
        [3.1415, 3.1415, 3.1415, 3.1415],
        [3.1415, 3.1415, 3.1415, 3.1415]]])

Dakle, morali smo prvo kreirati *Array* željenog oblika s jedinicama i onda ga pomnožiti sa željenom vrijednošću. Postoji  brži i učinkovitiji način. Metoda *full()* koja prima dva argumenta, jedan tuple s oblikom *Arraay* objekta, a drugi s brojem koji želimo imati u našem *Array* objektu.

In [45]:
np_pi = np.full((3, 4, 4), 7)
np_pi

array([[[7, 7, 7, 7],
        [7, 7, 7, 7],
        [7, 7, 7, 7],
        [7, 7, 7, 7]],

       [[7, 7, 7, 7],
        [7, 7, 7, 7],
        [7, 7, 7, 7],
        [7, 7, 7, 7]],

       [[7, 7, 7, 7],
        [7, 7, 7, 7],
        [7, 7, 7, 7],
        [7, 7, 7, 7]]])

Ponekad želimo imati *Array* u kojem su elementi poredani u sekvencu. Sjećate li se metode *range(start, stop, korak)*? NumPy radi s nizovima (*Arrays*) pa bi onda umjesto *array_range* trebala biti metoda naziva *arange*. Ostalo bi trebalo biti isto. Provjerimo.

In [48]:
np_arange = np.arange(5, 15)
np_arange = np.arange(5, 15, 3)
np_arange

array([ 5,  8, 11, 14])

Pomoću *arange()* metode generiramo jednodimenzionalni niz koji ima definiranu početnu vrijednost, krajnju (koja nije uključena) i korak između njih. NumPy nam omogućava kreiranje *Array* objekta koji ima početnu i krajnju vrijednost te broj elemenata između njih. Ova metoda zove se *linspace()*. Isprobajmo ju.

In [51]:
np_linspace = np.linspace(5, 15, 3)
np_linspace

array([ 5., 10., 15.])

In [52]:
np_linspace = np.linspace(5, 15, 5)
np_linspace

array([ 5. ,  7.5, 10. , 12.5, 15. ])

*arange* i *linspace* metode generiraju jednodimenzionalno polje.

**ZADATAK<br>
Generirajte nekoliko novih *Array* objekata pomoću *arange* i *linspace* metoda (neka ih bude minimalno 4) te im promijenite oblik po izboru.**

NumPy ima i generator nasumičnih brojeva unutar normalne distribucije (nula je sredina svih generiranih brojeva na ovaj način - brojevi se generiraju u pozitivnom i negativnom smjeru). Iskoristimo ga.

In [53]:
niz_brojeva = np.random.randn(4, 5)
niz_brojeva

array([[-0.50036545, -1.52337041,  0.36243233,  0.227734  , -1.10740138],
       [ 1.2287804 , -0.25083234, -0.57985369,  0.37205876, -0.75693202],
       [-1.94883679, -0.26495563,  1.46348674, -0.91380638, -1.0040406 ],
       [-1.29703793,  0.75003999,  0.45407813, -1.36304839,  2.24179479]])

Korisno. Ako nam je ovaj niz ipak nedovoljan, možemo ga jednostavno ponoviti onoliko puta koliko nam je potrebno. Za ponavljanje ćemo ipak koristiti čitljiviji niz kako bismo jednostavnije uočili obrasce ponavljanja.

Svakako napravite ponavljanje i s nasumično generiranim brojevima.

In [62]:
niz_brojeva = np.arange(1, 17).reshape(4, 4)
# niz_brojeva = niz_brojeva.repeat(4, axis=0) # Os 0 vertikalni smjer
niz_brojeva

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12],
       [13, 14, 15, 16]])

In [63]:
niz_brojeva = niz_brojeva.repeat(4, axis=1) # Os 1 horizontalni smjer
niz_brojeva

array([[ 1,  1,  1,  1,  2,  2,  2,  2,  3,  3,  3,  3,  4,  4,  4,  4],
       [ 5,  5,  5,  5,  6,  6,  6,  6,  7,  7,  7,  7,  8,  8,  8,  8],
       [ 9,  9,  9,  9, 10, 10, 10, 10, 11, 11, 11, 11, 12, 12, 12, 12],
       [13, 13, 13, 13, 14, 14, 14, 14, 15, 15, 15, 15, 16, 16, 16, 16]])

### *Slice* notacija

**ZADATAK<br>
Ako smo do sada mogli koristiti neke metode i funkcije vezane uz Python liste, provjerite možete li koristiti slice notaciju nad NumPy *Array* objektom. Podsjetnik: lista\[prvi_element : zadnji_element : korak\]. Sjetite se što znači negativni predznak za korak ili ako izostavite jedan od elemenata kao npr \[ : : \].**

Vjerujem da ste uspjeli jer se radi o jednodimenzionalnim nizovima. Međutim, kada radimo s višedimenzionalnim nizovima i ako koristimo *slice* notaciju, onda trebamo navesti i os na koju želimo primijeniti notaciju. Probajmo to napraviti s našim *tri_d Array* objektom.

In [65]:
tri_d = np.arange(1, 37).reshape(6, 6)
tri_d

array([[ 1,  2,  3,  4,  5,  6],
       [ 7,  8,  9, 10, 11, 12],
       [13, 14, 15, 16, 17, 18],
       [19, 20, 21, 22, 23, 24],
       [25, 26, 27, 28, 29, 30],
       [31, 32, 33, 34, 35, 36]])

Pokušajmo dohvatiti sve elemente od početnog, do elementa s indeksom 4 (nije uključen).

In [66]:
tri_d[ : 4]

array([[ 1,  2,  3,  4,  5,  6],
       [ 7,  8,  9, 10, 11, 12],
       [13, 14, 15, 16, 17, 18],
       [19, 20, 21, 22, 23, 24]])

Kao rezultat smo dobili prva četiri reda. Dakle, redove s indeksima 0, 1, 2 i 3. Što ako želimo dohvatiti prva četiri elementa drugog reda? Ako ponovimo gornju naredbu, onda ćemo dobiti sve iz prva četiri reda, a mi želimo samo neke elemente iz drugog reda.

NumPy ima rješenje i za to. U *slice* notaciji ćemo dodati još jedan argument na prvo mjesto koji će odrediti o kojem redu se radi, a nakon tog argumenta ćemo navesti koje elemente tog reda želimo.

In [69]:
tri_d[1, : 4]

array([ 7,  8,  9, 10])

Pokušajmo onda iskoristiti ovu notaciju da dohvatimo zadnja 4 elementa predzadnjeg reda.

In [72]:
tri_d[-2, -4 : ]

array([27, 28, 29, 30])

Vidjeli smo da je rad s redovima predefiniran. To znači kako nismo morali definirati o kojem se redu radi, nego je NumPy sam pretpostavio da želimo dohvatiti podatke iz redova. Međutim, što ako želimo dohvatiti podatke iz stupaca? Recimo, koji su podaci u drugom stupcu *tri_d* niza?

Odgovor na ovo pitanje je jako sličan odgovoru za redove, samo što argument koji navodimo za broj stupca, navodimo poslije *slice* notacije. Pokažimo na primjeru pa će biti jasnije:

In [73]:
tri_d[ : , 1]

array([ 2,  8, 14, 20, 26, 32])

Koja je vrijednost u zadnja četiri elementa četvrtog stupca?

In [74]:
tri_d[-4: , 3]

array([16, 22, 28, 34])

Dakle, *slice* notacija vrijedi i za NumPy nizove. Jednodimenzionalni nizovi su u stvari liste, kao što smo to naučili ranije, a ako želimo raditi s dvodimenzionalnim nizovima, onda slice notacija ima dodatne argumente. Prvi argument je indeks reda, nakon kojeg filtriramo pomoću *slice* notacije željene elemente. Ako želimo raditi sa stupcima, onda prvo navodimo *slice* notacijom filter elemenata, a nakon toga argument s indeksom stupca. Argumenti se, kao i kod funkcija, odvajaju zarezom.

Što ako želimo dohvatiti više redaka ili više stupaca? Ovdje ćemo morati koristiti dvostruku *slice* notaciju, odnosno kao da imamo dva argumenta. Prvi argument je zapisan u *slice* notaciji i odnosi se na redove, a drugi, također zapisan u *slice* notaciji, odnosi se na stupce. Zamislite kao da smo mišem selektirali dio tabele u Excel tabeli. Najbolje da to pokažemo na primjeru, u kojem ćemo dohvatiti prva četiri reda i zadnja četiri elementa tih prvih četiriju elementa.

In [75]:
tri_d

array([[ 1,  2,  3,  4,  5,  6],
       [ 7,  8,  9, 10, 11, 12],
       [13, 14, 15, 16, 17, 18],
       [19, 20, 21, 22, 23, 24],
       [25, 26, 27, 28, 29, 30],
       [31, 32, 33, 34, 35, 36]])

In [76]:
tri_d[ : 4, -4 : ]

array([[ 3,  4,  5,  6],
       [ 9, 10, 11, 12],
       [15, 16, 17, 18],
       [21, 22, 23, 24]])

**ZADATAK<br>
Kreirajte niz od 64 elemenata (početni i krajnji elementi neka budu proizvoljni). Prilagodite oblik niza tako da niz bude dvodimenzionalan.**

**Dohvatite samo PRVI RED, samo ZADNJI RED, samo PRVI STUPAC, samo ZADNJI STUPAC. Dohvatite centralne elemente niza (najmanja tabela u sredini niza).**

**Pokušajte sve to napraviti s trodimenzionalnim nizom. Koje zaključke ste donijeli?**

**Pomoć za *slice* notaciju**

<img src='./array-slice.png' alt="Array slice notacija" />

### Kreiranje niza pomoću funkcije

Prije nego krenemo dalje s jednostavnijim matematičkim operacijama, upoznajmo se s još jednom opcijom za kreiranje nizova. Niz možemo kreirati pomoću rezultata neke funkcije. Ovo nam može biti jako korisno.

Na primjer: imamo dva broja koje želimo transformirati tako da drugi broj pomnožimo s brojem 15 i to dodamo prvom broju. Rezultat ove funkcije ćemo iskoristiti kao ulazni parametar za kreiranje niza. Za to će nam koristiti metoda *fromfunction* koja kao prvi argument prima funkciju (dakle, koristimo *lambda* funkcije), drugi element je oblik, odnosno *tuple* s podacima za *shape* niza i treći opcionalni argument je tip podatka. Pokažimo na primjeru.

In [82]:
# niz = np.fromfunction(lambda x , y: 15 * y + x, (5, 8), dtype=np.int32)
# niz = np.fromfunction((lambda x , y: 15 * y + x), (5, 8), dtype=np.int32)
funkcija = lambda x , y, z: 15 * y + x - z
niz = np.fromfunction(funkcija, (5, 4, 4), dtype=np.int32)
niz

array([[[ 0, -1, -2, -3],
        [15, 14, 13, 12],
        [30, 29, 28, 27],
        [45, 44, 43, 42]],

       [[ 1,  0, -1, -2],
        [16, 15, 14, 13],
        [31, 30, 29, 28],
        [46, 45, 44, 43]],

       [[ 2,  1,  0, -1],
        [17, 16, 15, 14],
        [32, 31, 30, 29],
        [47, 46, 45, 44]],

       [[ 3,  2,  1,  0],
        [18, 17, 16, 15],
        [33, 32, 31, 30],
        [48, 47, 46, 45]],

       [[ 4,  3,  2,  1],
        [19, 18, 17, 16],
        [34, 33, 32, 31],
        [49, 48, 47, 46]]])

**PITANJE<br>
Objasnite, kako je moguće da funkcija vrati neki rezultat, ako u funkciju nismo proslijedili nikakve argumente?**

U funkciju smo indirektno proslijedili argumente na osnovu našeg *tuple* objekta koji definira oblik niza. Dakle, metoda *fromfunction* radi tako da poziva funkciju onoliko puta koliko niz ima elemenata na osnovu definiranog oblika. U našem slučaju to je 40. Prvi poziv funkcije je s argumentima funkcija (0, 0), drugi poziv je funkcija ( 0, 1), a zadnji poziv je funkcija (4, 7).

**Zadatak<br>
Prikažite tablicu množenja do 12 pomoću NumPy niza i funkcije.**

In [96]:
tablica_monzenja = np.fromfunction(lambda x, y: x * y , (12 + 1, 12 + 1), dtype=np.int32)
tablica_monzenja = tablica_monzenja[1 : , 1: ].copy() # Pomocu ovoga uklanjam redove i stupce u kojima su nule
tablica_monzenja

array([[  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12],
       [  2,   4,   6,   8,  10,  12,  14,  16,  18,  20,  22,  24],
       [  3,   6,   9,  12,  15,  18,  21,  24,  27,  30,  33,  36],
       [  4,   8,  12,  16,  20,  24,  28,  32,  36,  40,  44,  48],
       [  5,  10,  15,  20,  25,  30,  35,  40,  45,  50,  55,  60],
       [  6,  12,  18,  24,  30,  36,  42,  48,  54,  60,  66,  72],
       [  7,  14,  21,  28,  35,  42,  49,  56,  63,  70,  77,  84],
       [  8,  16,  24,  32,  40,  48,  56,  64,  72,  80,  88,  96],
       [  9,  18,  27,  36,  45,  54,  63,  72,  81,  90,  99, 108],
       [ 10,  20,  30,  40,  50,  60,  70,  80,  90, 100, 110, 120],
       [ 11,  22,  33,  44,  55,  66,  77,  88,  99, 110, 121, 132],
       [ 12,  24,  36,  48,  60,  72,  84,  96, 108, 120, 132, 144]])

## Kopiranje cijelog ili dijela niza

Sada smo naučili kako izdvojiti dio niza iz veće cjeline. Ako taj dio pohranimo u neku varijablu i dalje ćemo imati **referencu** na originalne podatke. Pokažimo to na primjeru tablice množenja:

In [94]:
tablica_monzenja_do100 = tablica_monzenja[ : 10, : 10 ]
tablica_monzenja_do100[ 4: 6 , 4 : 6 ] = 9999
print(tablica_monzenja)
print(tablica_monzenja_do100)

[[   1    2    3    4    5    6    7    8    9   10   11   12]
 [   2    4    6    8   10   12   14   16   18   20   22   24]
 [   3    6    9   12   15   18   21   24   27   30   33   36]
 [   4    8   12   16   20   24   28   32   36   40   44   48]
 [   5   10   15   20 9999 9999   35   40   45   50   55   60]
 [   6   12   18   24 9999 9999   42   48   54   60   66   72]
 [   7   14   21   28   35   42   49   56   63   70   77   84]
 [   8   16   24   32   40   48   56   64   72   80   88   96]
 [   9   18   27   36   45   54   63   72   81   90   99  108]
 [  10   20   30   40   50   60   70   80   90  100  110  120]
 [  11   22   33   44   55   66   77   88   99  110  121  132]
 [  12   24   36   48   60   72   84   96  108  120  132  144]]
[[   1    2    3    4    5    6    7    8    9   10]
 [   2    4    6    8   10   12   14   16   18   20]
 [   3    6    9   12   15   18   21   24   27   30]
 [   4    8   12   16   20   24   28   32   36   40]
 [   5   10   15   20 9999 9999

Kao što vidimo, promijenili smo **originalni skup podataka**. To treba izbjegavati, treba raditi na kopiji podataka. Za to imamo metodu *copy()*. Iskoristimo to na istom primjeru tablice množenja:

In [98]:
tablica_monzenja_do100 = tablica_monzenja[ : 10, : 10 ].copy()
tablica_monzenja_do100[ 4: 6 , 4 : 6 ] = 9999
print(tablica_monzenja)
print(tablica_monzenja_do100)

[[  1   2   3   4   5   6   7   8   9  10  11  12]
 [  2   4   6   8  10  12  14  16  18  20  22  24]
 [  3   6   9  12  15  18  21  24  27  30  33  36]
 [  4   8  12  16  20  24  28  32  36  40  44  48]
 [  5  10  15  20  25  30  35  40  45  50  55  60]
 [  6  12  18  24  30  36  42  48  54  60  66  72]
 [  7  14  21  28  35  42  49  56  63  70  77  84]
 [  8  16  24  32  40  48  56  64  72  80  88  96]
 [  9  18  27  36  45  54  63  72  81  90  99 108]
 [ 10  20  30  40  50  60  70  80  90 100 110 120]
 [ 11  22  33  44  55  66  77  88  99 110 121 132]
 [ 12  24  36  48  60  72  84  96 108 120 132 144]]
[[   1    2    3    4    5    6    7    8    9   10]
 [   2    4    6    8   10   12   14   16   18   20]
 [   3    6    9   12   15   18   21   24   27   30]
 [   4    8   12   16   20   24   28   32   36   40]
 [   5   10   15   20 9999 9999   35   40   45   50]
 [   6   12   18   24 9999 9999   42   48   54   60]
 [   7   14   21   28   35   42   49   56   63   70]
 [   8   16   24

### Povezivanje nizova

Osim što niz možemo kopirati pa tako kreirati novi niz, novi niz možemo kreirati i spajanjem dvaju ili više nizova u jedan pomoću *concatenate, vstack* i *hstack* metoda. Pokažimo to na primjeru.

In [99]:
niz_1 = np.array([[1, 2, 3], [4, 5, 6]])
print(niz_1)

print()

niz_2 = np.array([[7, 8, 9], [10, 11, 12]])
print(niz_2)

[[1 2 3]
 [4 5 6]]

[[ 7  8  9]
 [10 11 12]]


In [101]:
np.concatenate([niz_1, niz_2], axis=0)

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [102]:
np.concatenate([niz_1, niz_2], axis=1)

array([[ 1,  2,  3,  7,  8,  9],
       [ 4,  5,  6, 10, 11, 12]])

Možemo umjesto definiranja osi (*axis argument*) koristiti *vstack* (vertikalni smjer) ili *hstack* (horizontalni smjer)

In [104]:
np.vstack((niz_1, niz_2))

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [105]:
np.hstack((niz_1, niz_2))

array([[ 1,  2,  3,  7,  8,  9],
       [ 4,  5,  6, 10, 11, 12]])

### Razdvajanje nizova

Znamo kako izdvojiti dio niza u jednu novu zasebnu cjelinu, ali ponekad imamo potrebu niz podijeliti na više cjelina, odnosno nizova po nekom kriteriju. Za to koristimo metode *split, hsplit* i *vsplit*. Pokažimo na primjerima kako se koriste ove metode.

In [106]:
niz = np.arange(1, 17).reshape(4, 4)
niz

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12],
       [13, 14, 15, 16]])

In [108]:
np.hsplit(niz, 2)

[array([[ 1,  2],
        [ 5,  6],
        [ 9, 10],
        [13, 14]]),
 array([[ 3,  4],
        [ 7,  8],
        [11, 12],
        [15, 16]])]

In [109]:
np.vsplit(niz, 2)

[array([[1, 2, 3, 4],
        [5, 6, 7, 8]]),
 array([[ 9, 10, 11, 12],
        [13, 14, 15, 16]])]

Kao argument, na koliko dijelova možemo proslijediti jednodimenzionalni niz pa ćemo prvotni niz podijeliti na više različitih dijelova. Recimo, prvo ćemo izdvojiti prva tri stupca, zatim preostale elemente u novi niz sve do 6. stupca. Kako nemamo toliko, dobit ćemo i jedan prazan niz.

In [110]:
np.hsplit(niz, np.array([3, 6]))

[array([[ 1,  2,  3],
        [ 5,  6,  7],
        [ 9, 10, 11],
        [13, 14, 15]]),
 array([[ 4],
        [ 8],
        [12],
        [16]]),
 array([], shape=(4, 0), dtype=int32)]

Ako svaki novi dio pohranimo u neku novu varijablu, dobit ćemo nove *poglede* na originalni niz (sjetite se, nema kopiranja, nego i dalje zadržavamo reference pa ovo možemo smatrati kao pogled / *view* na taj dio niza.).

**Vježba<br>
Napravite nove kopije nizova.**

In [112]:
niz = np.arange(1, 65).reshape(8, 8)
np.hsplit(niz, np.array([3, 6]))

[array([[ 1,  2,  3],
        [ 9, 10, 11],
        [17, 18, 19],
        [25, 26, 27],
        [33, 34, 35],
        [41, 42, 43],
        [49, 50, 51],
        [57, 58, 59]]),
 array([[ 4,  5,  6],
        [12, 13, 14],
        [20, 21, 22],
        [28, 29, 30],
        [36, 37, 38],
        [44, 45, 46],
        [52, 53, 54],
        [60, 61, 62]]),
 array([[ 7,  8],
        [15, 16],
        [23, 24],
        [31, 32],
        [39, 40],
        [47, 48],
        [55, 56],
        [63, 64]])]

Ovo ćemo najlakše pojasniti na nekom većem nizu.

In [113]:
niz = np.arange(1, 65).reshape(8, 8)
np.hsplit(niz, np.array([3, 7]))

[array([[ 1,  2,  3],
        [ 9, 10, 11],
        [17, 18, 19],
        [25, 26, 27],
        [33, 34, 35],
        [41, 42, 43],
        [49, 50, 51],
        [57, 58, 59]]),
 array([[ 4,  5,  6,  7],
        [12, 13, 14, 15],
        [20, 21, 22, 23],
        [28, 29, 30, 31],
        [36, 37, 38, 39],
        [44, 45, 46, 47],
        [52, 53, 54, 55],
        [60, 61, 62, 63]]),
 array([[ 8],
        [16],
        [24],
        [32],
        [40],
        [48],
        [56],
        [64]])]

In [115]:
np.split?

Signature:       np.split(ary, indices_or_sections, axis=0)
Call signature:  np.split(*args, **kwargs)
Type:            _ArrayFunctionDispatcher
String form:     <function split at 0x0000025EA8ACF2E0>
File:            c:\repos\py-jupyter-nb\venv\lib\site-packages\numpy\lib\shape_base.py
Docstring:      
Split an array into multiple sub-arrays as views into `ary`.

Parameters
----------
ary : ndarray
    Array to be divided into sub-arrays.
indices_or_sections : int or 1-D array
    If `indices_or_sections` is an integer, N, the array will be divided
    into N equal arrays along `axis`.  If such a split is not possible,
    an error is raised.

    If `indices_or_sections` is a 1-D array of sorted integers, the entries
    indicate where along `axis` the array is split.  For example,
    ``[2, 3]`` would, for ``axis=0``, result in

      - ary[:2]
      - ary[2:3]
      - ary[3:]

    If an index exceeds the dimension of the array along `axis`,
    an empty sub-array is returned corres

## Matematičke operacije s nizovima

NumPy je nastao kako bismo brzo i jednostavno izvršavali kompleksne matematičke operacije nad velikom količinom podatka. Da bismo došli do kompleksnih matematičkih operacija, krenut ćemo s jednostavnijim kao što su osnovne matematičke operacije:
- zbrajanje i oduzimanje
- množenje i dijeljenje
- potencije.

In [116]:
niz_a = np.arange(1, 65).reshape(8, 8)
niz_b = np.arange(101, 165).reshape(8, 8)

niz_a + niz_b

array([[102, 104, 106, 108, 110, 112, 114, 116],
       [118, 120, 122, 124, 126, 128, 130, 132],
       [134, 136, 138, 140, 142, 144, 146, 148],
       [150, 152, 154, 156, 158, 160, 162, 164],
       [166, 168, 170, 172, 174, 176, 178, 180],
       [182, 184, 186, 188, 190, 192, 194, 196],
       [198, 200, 202, 204, 206, 208, 210, 212],
       [214, 216, 218, 220, 222, 224, 226, 228]])

Ili pomoću funkcije metode *add()*:

In [117]:
np.add(niz_a, niz_b)

array([[102, 104, 106, 108, 110, 112, 114, 116],
       [118, 120, 122, 124, 126, 128, 130, 132],
       [134, 136, 138, 140, 142, 144, 146, 148],
       [150, 152, 154, 156, 158, 160, 162, 164],
       [166, 168, 170, 172, 174, 176, 178, 180],
       [182, 184, 186, 188, 190, 192, 194, 196],
       [198, 200, 202, 204, 206, 208, 210, 212],
       [214, 216, 218, 220, 222, 224, 226, 228]])

Analogno prethodnom primjeru, oduzimanje bi bilo pomoću znaka - ili metode *subtract()*.

In [118]:
niz_a - niz_b

array([[-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100]])

In [119]:
np.subtract(niz_a, niz_b)

array([[-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100],
       [-100, -100, -100, -100, -100, -100, -100, -100]])

In [120]:
np.multiply(niz_a, niz_b)

array([[  101,   204,   309,   416,   525,   636,   749,   864],
       [  981,  1100,  1221,  1344,  1469,  1596,  1725,  1856],
       [ 1989,  2124,  2261,  2400,  2541,  2684,  2829,  2976],
       [ 3125,  3276,  3429,  3584,  3741,  3900,  4061,  4224],
       [ 4389,  4556,  4725,  4896,  5069,  5244,  5421,  5600],
       [ 5781,  5964,  6149,  6336,  6525,  6716,  6909,  7104],
       [ 7301,  7500,  7701,  7904,  8109,  8316,  8525,  8736],
       [ 8949,  9164,  9381,  9600,  9821, 10044, 10269, 10496]])

In [121]:
np.divide(niz_a, niz_b)

array([[0.00990099, 0.01960784, 0.02912621, 0.03846154, 0.04761905,
        0.05660377, 0.06542056, 0.07407407],
       [0.08256881, 0.09090909, 0.0990991 , 0.10714286, 0.11504425,
        0.12280702, 0.13043478, 0.13793103],
       [0.14529915, 0.15254237, 0.15966387, 0.16666667, 0.17355372,
        0.18032787, 0.18699187, 0.19354839],
       [0.2       , 0.20634921, 0.21259843, 0.21875   , 0.2248062 ,
        0.23076923, 0.23664122, 0.24242424],
       [0.2481203 , 0.25373134, 0.25925926, 0.26470588, 0.27007299,
        0.27536232, 0.28057554, 0.28571429],
       [0.29078014, 0.29577465, 0.3006993 , 0.30555556, 0.31034483,
        0.31506849, 0.31972789, 0.32432432],
       [0.32885906, 0.33333333, 0.33774834, 0.34210526, 0.34640523,
        0.35064935, 0.35483871, 0.35897436],
       [0.36305732, 0.36708861, 0.37106918, 0.375     , 0.37888199,
        0.38271605, 0.38650307, 0.3902439 ]])

In [123]:
np.power(niz_a, 3)

array([[     1,      8,     27,     64,    125,    216,    343,    512],
       [   729,   1000,   1331,   1728,   2197,   2744,   3375,   4096],
       [  4913,   5832,   6859,   8000,   9261,  10648,  12167,  13824],
       [ 15625,  17576,  19683,  21952,  24389,  27000,  29791,  32768],
       [ 35937,  39304,  42875,  46656,  50653,  54872,  59319,  64000],
       [ 68921,  74088,  79507,  85184,  91125,  97336, 103823, 110592],
       [117649, 125000, 132651, 140608, 148877, 157464, 166375, 175616],
       [185193, 195112, 205379, 216000, 226981, 238328, 250047, 262144]],
      dtype=int32)

**PITANJE <br>
Zbog čega smo kao rezultat oduzimanja dobili sve iste brojeve?**

Odgovor: zato jer je razlika elemenata po pozicijama uvijek identična. Da smo kreirali drugačije nizove, razlika bi bila drugačija.

**Primjere za množenje i dijeljenje, pokušajte napraviti sami.<br>
**Pomoć: koristite engleske nazive za množenje i dijeljenje, odnosno *multiply* i *divide.***

### Ugrađene matematičke funkcije u NumPy biblioteci

NumPy ima jako puno ugrađenih funkcija. Da nabrojimo samo neke:
- trigonometrijske funkcije: *sin(), cos()* i *tan()*
- funkcije za korjenovanje *sqrt()* te *exp()* funkciju za računanje potencija nekog broja
- *log(), log2()* i *log10()* za računanje logaritama s bazom *e*, 2 ili 10
- *round()* za zaokruživanje na određeni broj decimalnih mjesta.

**Generirajte niz s nasumičnim brojevima proizvoljnog oblika (minimalno 4,4) te nad tim nizom provjerite kako rade navedene funkcije.**

### Statističke operacije u NumPy biblioteci
- *np.mean* - prosjek svih vrijednosti u nizu
- *np.std* - standardna devijacija
- *np.sum* - suma svih elemenata, metoda koristi argument *axis*
- *np.prod* - produkt svih elemenata
- *np.min, np.max* - minimalna/maksimalna vrijednost u nizu
- *np.argmin, np.argmax* - indeks elementa koji ima minimalnu/maksimalnu vrijednost
- *np.all* - vraća True ako su svi elementi nonzero (elementi koji imaju vrijednost nula ili nemaju definiranu vrijednost)
- *np.any* - vraća True ako je bilo koji element *nonzero*

Generirajmo niz nasumičnih brojeva i provjerimo neke od gore navedenih statističkih operacija.

In [127]:
niz = np.random.randn(4, 6)
niz = np.round(niz, 3)

print(niz)
print(np.mean(niz))
print(np.std(niz))
print(np.sum(niz))

[[-1.43   1.164 -1.587  0.318  2.121 -0.187]
 [-0.019  1.857  1.16   2.259 -0.585  0.155]
 [ 0.152  0.116  1.348  0.834 -0.012 -1.394]
 [ 0.623  0.154  0.532  0.81  -0.939  0.722]]
0.3405
1.02633750134479
8.172


**Vježba<br>
Dovršite primjere za preostale statističke funkcije tako da za svaku kreirate novi niz.**

### Funkcije filtriranja

NumPy ima ugrađenih nekoliko funkcija za filtriranje i obradu vrijednosti elementa u nizovima.
Prva s kojom ćemo se upoznati je *where()*. Ova funkcija prima prvi argument kao uvjet za filtriranje, a druga dva su operacije nad vrijednostima elemenata niza. Ako je rezultat uvjeta iz prvog argumenta točan, onda se primjenjuje drugi argument, a ako nije točan, onda se primjenjuje treći argument.

Generirajmo niz od 9 elemenata od -4 do +4 i svaki element koji je manji od nula podignimo na kvadrat, a ako je veći od nula, podignimo na treću potenciju.

In [131]:
niz = np.linspace(-10, 10, 11)
niz

array([-10.,  -8.,  -6.,  -4.,  -2.,   0.,   2.,   4.,   6.,   8.,  10.])

In [132]:
np.where?

Call signature:  np.where(*args, **kwargs)
Type:            _ArrayFunctionDispatcher
String form:     <built-in function where>
Docstring:      
where(condition, [x, y], /)

Return elements chosen from `x` or `y` depending on `condition`.

.. note::
    When only `condition` is provided, this function is a shorthand for
    ``np.asarray(condition).nonzero()``. Using `nonzero` directly should be
    preferred, as it behaves correctly for subclasses. The rest of this
    documentation covers only the case where all three arguments are
    provided.

Parameters
----------
condition : array_like, bool
    Where True, yield `x`, otherwise yield `y`.
x, y : array_like
    Values from which to choose. `x`, `y` and `condition` need to be
    broadcastable to some shape.

Returns
-------
out : ndarray
    An array with elements from `x` where `condition` is True, and elements
    from `y` elsewhere.

See Also
--------
choose
nonzero : The function that is called when x and y are omitted

Notes


In [135]:
np.where(niz < 0, niz**2, niz/2)

array([100.,  64.,  36.,  16.,   4.,   0.,   1.,   2.,   3.,   4.,   5.])

Druga jako slična funkcija je *select()*. Ova funkcija prima listu uvjeta kao prvi argument i listu operacija nad elementima koji zadovoljavaju uvjete iz liste. Nad prethodno kreiranim nizom, pronađimo elemente koji su:
- manji od -1 i ako ih pronađemo podignimo ih na kvadrat
- manji od 2 i ako ih pronađemo podignimo ih na treću potenciju
- veći i jednaki od 2 te ako ih pronađemo podignimo ih na četvrtu potenciju.

In [136]:
np.select?

Signature:       np.select(condlist, choicelist, default=0)
Call signature:  np.select(*args, **kwargs)
Type:            _ArrayFunctionDispatcher
String form:     <function select at 0x0000025EA8ABC040>
File:            c:\repos\py-jupyter-nb\venv\lib\site-packages\numpy\lib\function_base.py
Docstring:      
Return an array drawn from elements in choicelist, depending on conditions.

Parameters
----------
condlist : list of bool ndarrays
    The list of conditions which determine from which array in `choicelist`
    the output elements are taken. When multiple conditions are satisfied,
    the first one encountered in `condlist` is used.
choicelist : list of ndarrays
    The list of arrays from which the output elements are taken. It has
    to be of the same length as `condlist`.
default : scalar, optional
    The element inserted in `output` when all conditions evaluate to False.

Returns
-------
output : ndarray
    The output at position m is the m-th element of the array in
    `c

In [138]:
np.select([niz < -1, niz < 2, niz >= 2], [niz**2, niz**3, niz / 5])

array([100. ,  64. ,  36. ,  16. ,   4. ,   0. ,   0.4,   0.8,   1.2,
         1.6,   2. ])

**np.unique** je funkcija koja vraća **jedinstvene** elemente niza

In [141]:
niz = np.random.randint(5, size=20)
np.unique(niz)

array([0, 1, 2, 3, 4])

Prije nego krenemo s naprednom primjenom NumPy biblioteke, upoznajmo se s još nekoliko funkcija vezanih uz nizove:
- *np.fliplr* - sortira redove obrnutim redoslijedom - *fliplr = Flip L(eft)R(ight)*
- *np.flipud* - sortira stupce obrnutim redoslijedom - *flipud = Flip U(p)D(own)*
- *np.sort* - sortira niz po zadanim osima
- *np.ndarray.sort* - sortira niz tako da ga modificira.

In [142]:
niz = np.arange(1, 65).reshape(8, 8)
niz

array([[ 1,  2,  3,  4,  5,  6,  7,  8],
       [ 9, 10, 11, 12, 13, 14, 15, 16],
       [17, 18, 19, 20, 21, 22, 23, 24],
       [25, 26, 27, 28, 29, 30, 31, 32],
       [33, 34, 35, 36, 37, 38, 39, 40],
       [41, 42, 43, 44, 45, 46, 47, 48],
       [49, 50, 51, 52, 53, 54, 55, 56],
       [57, 58, 59, 60, 61, 62, 63, 64]])

In [143]:
np.fliplr(niz)

array([[ 8,  7,  6,  5,  4,  3,  2,  1],
       [16, 15, 14, 13, 12, 11, 10,  9],
       [24, 23, 22, 21, 20, 19, 18, 17],
       [32, 31, 30, 29, 28, 27, 26, 25],
       [40, 39, 38, 37, 36, 35, 34, 33],
       [48, 47, 46, 45, 44, 43, 42, 41],
       [56, 55, 54, 53, 52, 51, 50, 49],
       [64, 63, 62, 61, 60, 59, 58, 57]])

In [144]:
np.flipud(niz)

array([[57, 58, 59, 60, 61, 62, 63, 64],
       [49, 50, 51, 52, 53, 54, 55, 56],
       [41, 42, 43, 44, 45, 46, 47, 48],
       [33, 34, 35, 36, 37, 38, 39, 40],
       [25, 26, 27, 28, 29, 30, 31, 32],
       [17, 18, 19, 20, 21, 22, 23, 24],
       [ 9, 10, 11, 12, 13, 14, 15, 16],
       [ 1,  2,  3,  4,  5,  6,  7,  8]])

In [146]:
np.sort(niz, axis=0)

array([[ 1,  2,  3,  4,  5,  6,  7,  8],
       [ 9, 10, 11, 12, 13, 14, 15, 16],
       [17, 18, 19, 20, 21, 22, 23, 24],
       [25, 26, 27, 28, 29, 30, 31, 32],
       [33, 34, 35, 36, 37, 38, 39, 40],
       [41, 42, 43, 44, 45, 46, 47, 48],
       [49, 50, 51, 52, 53, 54, 55, 56],
       [57, 58, 59, 60, 61, 62, 63, 64]])

In [147]:
np.sort(niz, axis=1)

array([[ 1,  2,  3,  4,  5,  6,  7,  8],
       [ 9, 10, 11, 12, 13, 14, 15, 16],
       [17, 18, 19, 20, 21, 22, 23, 24],
       [25, 26, 27, 28, 29, 30, 31, 32],
       [33, 34, 35, 36, 37, 38, 39, 40],
       [41, 42, 43, 44, 45, 46, 47, 48],
       [49, 50, 51, 52, 53, 54, 55, 56],
       [57, 58, 59, 60, 61, 62, 63, 64]])

U sljedećem poglavlju ćemo se baviti analizom stvarnih podataka, u kojemu ćemo koristiti prethodno naučene biblioteke Pandas i matplotlib. matplotlib i Pandas su biblioteke razvijene na osnovi NumPy biblioteke te ćemo ih koristiti ovisno o njihovim najjačim stranama.

### Priča o najpoznatijoj matrici na svijetu - šahovskoj ploči

**Šah**

Najpoznatija matrica je ploča za jednu od najstarijih igara - šah. To je matrica 8x8.
Legenda o tome kako je igra dobila ime šah je jako zanimljiva. Šah (perzijski naziv za vladara) je gledao kako njegove sluge od putujućih trgovaca kupuju namirnice, tkaninu, začine i sl. Tako je ugledao dvojicu trgovaca kako igraju neku čudnu igru. Poslao je slugu da dovede trgovca koji je donio tu igru u njegovo kraljevstvo. Šah (vladar) je zatražio od trgovca da ga nauči igrati. Kada je naučio kako se igra, šah se toliko oduševio igrom da je odlučio bogato nagraditi trgovca. Rekao mu je neka kaže što želi kao nagradu za ovu igru. Trgovac je šahu, nakon što je naveo sve one slatkorječive riječi vladarima, kao što su Vaša visosti i sl., rekao:

*Želim pšenicu. Na prvo polje stavite jedno zrno pšenice. Na drugo dva zrna, na treće četiri, na četvrto dva puta više od prethodnog ... i tako do posljednjeg 64.-og polja.*

Šah se jako uvrijedio jer je očekivao da će trgovac tražiti zlato, drago kamenje, skupe začine pa ga je potjerao u vrt rekavši mu da će svoju vreću pšenice dobiti za ručak. Slugama je naredio neka ga isplate i to da nikada više ne želi vidjeti tog trgovca u njegovom kraljevstvu.

**PITANJE<br>
Koliko je trgovac trebao dobiti pšenice? Napišite program koji će to izračunati. Ako znate ovu priču, svejedno dozvolite drugim kolegama neka sami pokušaju doći do odgovora i svakako napišite program.**

In [153]:
broj_zrna_psenice = 0

for i in range(64):
    if i == 0:
        broj_zrna_psenice += 1
    if i == 1:
        broj_zrna_psenice += 2
    else:
        broj_zrna_psenice *= 2

print(broj_zrna_psenice)
print(round(broj_zrna_psenice / (40 * 100 * 100 * 1000000), 2), 'km2')

18446744073709551616
46116860.18 km2


Rezultat je: 18.446.744.073.709.551.616,00!<br>
Zna li netko pročitati ovaj broj?

Prošlo je nekoliko dana i šah je u vrtu vidio trgovca. Zvao je sluge i pitao što on još radi u njegovom kraljevstvu, odnosno zar nije isplaćen. Sluge su mu rekle da su njegovi matematičari izračunali:
- da svaki klas pšenice ima 40 zrna (četiri puta po 10 zrna - pogledajte sliku na internetu)
- da na jednom kvadratnom centimetru raste jedna stabljika pšenice s jednim klasom, na jednom kvadratnom metru ima 100 x 100 = 10.000,00 stabljika pšenica, dakle 400.000,00 zrna po kvadratnom metru
- da jedan kvadratni kilometar ima 1.000.000,00 kvadratnih metara pa na jednom kvadratnom kilometru ima 400.000.000.000,00 zrna pšenice.

Potrebno je posijati 46.116.860,18 (46 milijuna, 116 tisuća i 860,18) kvadratnih kilometara pšenicom da bi se trgovac isplatio. Usporedbe radi, ukupna površina kontinenta Azije je 44.580.000,00 kvadratnih kilometara.

Shvativši da ne može isplatiti trgovca, šah priđe trgovcu te mu se ispriča što je morao toliko čekati, ali se čekanje isplatilo. Pšenica je spremna za njegovu isplatu, ali šah ne bi volio da se trgovac osjeti prevarenim pa neka on sam izbroji svoju nagradu, a šahove sluge će biti uz njega i biti mu na usluzi dok on broji.

Trgovac je znao što ga čeka pa je odgovorio šahu da mu se zahvaljuje na nagradi, da mu nagrada sada nije važna te moli šaha da mu dozvoli da odmah krene kući jer je jutros dobio glas da mu je otac smrtno bolestan te bi se želio oprostiti od njega dok još može. Šah je dozvolio odlazak trgovca te mu dao dva konja puna namirnica potrebnih za put i još nešto zlatnika sa svrhom da požuri kući ocu, ali i da svijetom širi glas o milostivosti šaha.

*Napomena: legenda može odstupati od neke pisane verzije jer je ispričana po sjećanju.*